llama_stack/providers/impls/meta_reference/datasetio/datasetio.py

```diff
@@ -0,0 +1,126 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the terms described in the LICENSE file in
+# the root directory of this source tree.
+from typing import List, Optional
+
+import pandas
+
+from llama_models.llama3.api.datatypes import *  # noqa: F403
+
+from llama_stack.apis.datasetio import *  # noqa: F403
+from dataclasses import dataclass
+
+from llama_stack.providers.datatypes import DatasetsProtocolPrivate
+from llama_stack.providers.utils.datasetio.dataset_utils import BaseDataset
+
+from .config import MetaReferenceDatasetIOConfig
+
+
+@dataclass
+class DatasetInfo:
+    dataset_def: DatasetDef
+    dataset_impl: BaseDataset
+    next_page_token: Optional[int] = None
+
+
+class CustomDataset(BaseDataset):
+    def __init__(self, dataset_def: DatasetDef, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self.dataset_def = dataset_def
+        # TODO: validate dataset_def against schema
+        self.df = None
+        self.load()
+
+    def __len__(self) -> int:
```

Comment:
why do we need to pretend this dataset to be an array? seems like adding a `num_rows()` method is clearer? 

also please don't `load()` automatically -- anybody who is calling `num_rows()` should have called `load()` first

Reply:
Was following the dataset interface from [PyTorch](https://pytorch.org/tutorials/beginner/basics/data_tutorial.html)/[Huggingface](https://huggingface.co/docs/datasets/en/quickstart) that allows array-like operations. Could change it to num_rows instead.

Reply:
Ah I see. Let's keep it!